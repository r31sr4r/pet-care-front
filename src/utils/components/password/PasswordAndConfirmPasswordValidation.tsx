import { useState, useEffect } from 'react';
import Grid from '@mui/material/Grid';
import TextField from '@mui/material/TextField';
import { User } from '../../../types/User';

type Props = {
	user: User;
	handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
}

export function PasswordAndConfirmPasswordValidation({
	user,
	handleChange,
}: Props) {
	const [passwordError, setPasswordError] = useState(false);
	const [confirmPasswordError, setConfirmPasswordError] = useState(false);
	const [passwordErrorMessages, setPasswordErrorMessages] = useState('');

	const handleValidation = () => {
		const upperCaseRegex = /(?=.*?[A-Z])/;
		const lowerCaseRegex = /(?=.*?[a-z])/;
		const numberRegex = /(?=.*?\d)/;
		const specialCharacterRegex = /(?=.*?[#?!@$%^&*-])/;
		const minLengthRegex = /(?=.{6,})/;

		const upperCaseValidation = upperCaseRegex.test(user.password);
		const lowerCaseValidation = lowerCaseRegex.test(user.password);
		const numberValidation = numberRegex.test(user.password);
		const specialCharacterValidation = specialCharacterRegex.test(
			user.password
		);
		const minLengthValidation = minLengthRegex.test(user.password);

		const passwordValidation: boolean =
			upperCaseValidation &&
			lowerCaseValidation &&
			(numberValidation || specialCharacterValidation) &&
			minLengthValidation;

		const numberAndSpecialCharacterValidation: boolean =
			numberValidation || specialCharacterValidation;

		const passwordErrorMessages = `Sua senha precisar ter pelo menos:
            ${upperCaseValidation ? '✔️' : '❌'} uma letra maíuscula
            ${lowerCaseValidation ? '✔️' : '❌'} uma letra minúscula
            ${
				numberAndSpecialCharacterValidation ? '✔️' : '❌'
			} um número ou caracter especial            
            ${minLengthValidation ? '✔️' : '❌'} no mínimo 6 caracteres`;

		if (!passwordValidation) {
			setPasswordError(true);
			setPasswordErrorMessages(passwordErrorMessages);
		} else {
			setPasswordError(false);
			setPasswordErrorMessages('');
		}

		if (user.password !== user.confirm_password) {
			setPasswordError(true);
			setConfirmPasswordError(true);
		} else {
			setPasswordError(false);
			setConfirmPasswordError(false);
		}
	};

	useEffect(() => {
		handleValidation();
	}, [user.password, user.confirm_password]);

	return (
		<>
			<Grid item xs={12}>
				<TextField
					required
					fullWidth
					name="password"
					label="Senha"
					type="password"
					id="password"
					autoComplete="new-password"
					value={user.password}
					onChange={handleChange}
					error={passwordError}
					helperText={passwordError ? passwordErrorMessages : ''}
				/>
			</Grid>
			<Grid item xs={12}>
				<TextField
					required
					fullWidth
					name="confirm_password"
					label="Confirmar senha"
					type="password"
					id="confirm_password"
					autoComplete="new-password"
					value={user.confirm_password}
					onChange={handleChange}
					error={confirmPasswordError}
					helperText={
						confirmPasswordError ? 'As senhas não conferem' : ''
					}
				/>
			</Grid>
		</>
	);
}
