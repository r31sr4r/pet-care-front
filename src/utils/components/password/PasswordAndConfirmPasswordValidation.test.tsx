import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { PasswordAndConfirmPasswordValidation } from './PasswordAndConfirmPasswordValidation';

const mockUser = {
	id: '1',
	name: 'User Test',
	email: 'user@mail.com',
	password: 'Password123@',
	old_password: null,
	confirm_password: 'Password123@',
	group: 'USER',
	role: 'USER',
	is_active: true,
	created_at: '2021-01-01T00:00:00.000Z',
};

const Props = {
	user: mockUser,
	handleChange: () => {},
};

describe('PasswordAndConfirmPasswordValidation', () => {
	it('should render successfully', () => {
		const { asFragment } = render(
			<PasswordAndConfirmPasswordValidation {...Props} />,
			{
				wrapper: BrowserRouter,
			}
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('should render with a invalid password', () => {
		const { asFragment } = render(
			<PasswordAndConfirmPasswordValidation
				{...Props}
				user={{ ...mockUser, password: 'pa' }}
			/>,
			{
				wrapper: BrowserRouter,
			}
		);
		expect(asFragment()).toMatchSnapshot();
	});

    it('should render with other invalid password', () => {
		const { asFragment } = render(
			<PasswordAndConfirmPasswordValidation
				{...Props}
				user={{ ...mockUser, password: 'SW' }}
			/>,
			{
				wrapper: BrowserRouter,
			}
		);
		expect(asFragment()).toMatchSnapshot();
	});


    it('should render with a valid password', () => {
        const { asFragment } = render(
            <PasswordAndConfirmPasswordValidation
                {...Props}
                user={{ ...mockUser, password: 'Password@' }}
            />,
            {
                wrapper: BrowserRouter,
            }
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
