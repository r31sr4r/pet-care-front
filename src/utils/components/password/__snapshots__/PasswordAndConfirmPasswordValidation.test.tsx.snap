// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordAndConfirmPasswordValidation should render successfully 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="password"
        id="password-label"
      >
        Senha
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="new-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="password"
          name="password"
          required=""
          type="password"
          value="Password123@"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Senha *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="confirm_password"
        id="confirm_password-label"
      >
        Confirmar senha
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="false"
          autocomplete="new-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="confirm_password"
          name="confirm_password"
          required=""
          type="password"
          value="Password123@"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Confirmar senha *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PasswordAndConfirmPasswordValidation should render with a invalid password 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="password"
        id="password-label"
      >
        Senha
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-describedby="password-helper-text"
          aria-invalid="true"
          autocomplete="new-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="password"
          name="password"
          required=""
          type="password"
          value="pa"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Senha *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled Mui-required css-1wc848c-MuiFormHelperText-root"
        id="password-helper-text"
      >
        Sua senha precisar ter pelo menos:
            ❌ uma letra maíuscula
            ✔️ uma letra minúscula
            ❌ um número ou caracter especial            
            ❌ no mínimo 6 caracteres
      </p>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="confirm_password"
        id="confirm_password-label"
      >
        Confirmar senha
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-describedby="confirm_password-helper-text"
          aria-invalid="true"
          autocomplete="new-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="confirm_password"
          name="confirm_password"
          required=""
          type="password"
          value="Password123@"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Confirmar senha *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled Mui-required css-1wc848c-MuiFormHelperText-root"
        id="confirm_password-helper-text"
      >
        As senhas não conferem
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PasswordAndConfirmPasswordValidation should render with a valid password 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="password"
        id="password-label"
      >
        Senha
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-invalid="true"
          autocomplete="new-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="password"
          name="password"
          required=""
          type="password"
          value="Password@"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Senha *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="confirm_password"
        id="confirm_password-label"
      >
        Confirmar senha
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-describedby="confirm_password-helper-text"
          aria-invalid="true"
          autocomplete="new-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="confirm_password"
          name="confirm_password"
          required=""
          type="password"
          value="Password123@"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Confirmar senha *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled Mui-required css-1wc848c-MuiFormHelperText-root"
        id="confirm_password-helper-text"
      >
        As senhas não conferem
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PasswordAndConfirmPasswordValidation should render with other invalid password 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="password"
        id="password-label"
      >
        Senha
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-describedby="password-helper-text"
          aria-invalid="true"
          autocomplete="new-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="password"
          name="password"
          required=""
          type="password"
          value="SW"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Senha *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled Mui-required css-1wc848c-MuiFormHelperText-root"
        id="password-helper-text"
      >
        Sua senha precisar ter pelo menos:
            ✔️ uma letra maíuscula
            ❌ uma letra minúscula
            ❌ um número ou caracter especial            
            ❌ no mínimo 6 caracteres
      </p>
    </div>
  </div>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-13i4rnv-MuiGrid-root"
  >
    <div
      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-required css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
        data-shrink="true"
        for="confirm_password"
        id="confirm_password-label"
      >
        Confirmar senha
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk Mui-error css-wgai2y-MuiFormLabel-asterisk"
        >
           *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
      >
        <input
          aria-describedby="confirm_password-helper-text"
          aria-invalid="true"
          autocomplete="new-password"
          class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
          id="confirm_password"
          name="confirm_password"
          required=""
          type="password"
          value="Password123@"
        />
        <fieldset
          aria-hidden="true"
          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="css-14lo706"
          >
            <span>
              Confirmar senha *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled Mui-required css-1wc848c-MuiFormHelperText-root"
        id="confirm_password-helper-text"
      >
        As senhas não conferem
      </p>
    </div>
  </div>
</DocumentFragment>
`;
